src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1286,20 +1286,26 @@ private async ValueTask<(Stream?, TransportContext?, HttpResponseMessage?)> Conn
             }
         }
 
-        private static readonly SocketsConnectionFactory s_defaultConnectionFactory = new SocketsConnectionFactory(SocketType.Stream, ProtocolType.Tcp);
+        private static readonly SocketsConnectionFactory s_connectionFactoryInstance = new SocketsConnectionFactory(SocketType.Stream, ProtocolType.Tcp);
+        private static readonly Func<DnsEndPoint, HttpRequestMessage, CancellationToken, ValueTask<Stream>> s_defaultConnectionCallback = (ep, r, token) => s_connectionFactoryInstance.ConnectAsync(ep, token);
 
         private ValueTask<Stream> ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
         {
             if (async)
             {
-                SocketsConnectionFactory connectionFactory = s_defaultConnectionFactory;
+                Func<DnsEndPoint, HttpRequestMessage, CancellationToken, ValueTask<Stream>> connectionCallback = Settings._connectionCallback ?? s_defaultConnectionCallback;
 
                 var endPoint = new DnsEndPoint(host, port);
-                return ConnectHelper.ConnectAsync(connectionFactory, endPoint, cancellationToken);
+                return ConnectHelper.ConnectAsync(connectionCallback, endPoint, initialRequest, cancellationToken);
             }
 
             // Synchronous path.
 
+            if (Settings._connectionCallback is not null)
+            {
+                throw new InvalidOperationException(SR.net_http_sync_operations_not_allowed_with_connect_callback);
```

Comment:
We're throwing NSE in case certain operation doesn't support sync, e.g.: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L351
If you think IOE is more suitable, we should probably change it in places I introduced the NSE.

Reply:
Nope, I'm fine with NSE. Will change.